File: nodejs/src/capabilities/v1_1/layout.ts
/**
 * File_Version 1.1 on-disk HDF5 layout (Node).
 *
 * Deliberately independent of the previous version's layout module: nothing
 * here imports from or refers to it, so that module can change or be removed
 * later without affecting this one.
 */
export const FILE_VERSION = '1.1';

export const ROOT_MACHINE = 'Machine';
export const ROOT_OPTICAL_TRAINS = 'Machine/Optical_Trains';
export const ROOT_EXTENSIONS = 'Extensions';
export const GROUP_CLEARBOX = 'Extensions/ClearBox';
export const ROOT_OPCUA = 'Extensions/TM_OPCUA';
export const OPCUA_CLIENT = 'Extensions/TM_OPCUA/Client';
export const OPCUA_PIPE = 'Extensions/TM_OPCUA/Pipe';
export const OPCUA_TRIGGERS = 'Extensions/TM_OPCUA/Triggers';

export const TRAIN_ID_PREFIX = 'Optical_Train_';
export const GROUP_SCANNER = 'Scanner';
export const GROUP_LIGHT_SOURCE = 'Light_Source';
export const GROUP_COLLIMATOR = 'Collimator';
export const GROUP_SCANNER_CARD = 'Scanner_Card';
export const GROUP_POWER_CHARACTERIZATION = 'Power_Characterization';
export const DS_DERIVATION_EQUATION_CONSTANTS = 'Derivation_Equation_Constants';
export const DS_CHARACTERIZATION_POINTS = 'Characterization_Points';
export const DS_CORRECTION_DATA = 'Correction_Data';
export const DS_INVERSE_CORRECTION_DATA = 'Inverse_Correction_Data';
export const DS_SCAN_FIELD_CORRECTION_FILE = 'scan_field_correction_file';

export function trainId(index: number): string {
  return `${TRAIN_ID_PREFIX}${String(index + 1).padStart(2, '0')}`;
}

export function trainIds(keys: string[]): string[] {
  return keys.filter((k) => k.startsWith(TRAIN_ID_PREFIX)).sort();
}

export function trainPathById(tid: string): string {
  return `${ROOT_OPTICAL_TRAINS}/${tid}`;
}

export function trainPath(index: number): string {
  return trainPathById(trainId(index));
}

export function scannerPathById(tid: string): string {
  return `${trainPathById(tid)}/${GROUP_SCANNER}`;
}

export function lightSourcePathById(tid: string): string {
  return `${trainPathById(tid)}/${GROUP_LIGHT_SOURCE}`;
}

export function collimatorPathById(tid: string): string {
  return `${trainPathById(tid)}/${GROUP_COLLIMATOR}`;
}

export function scannerCardPathById(tid: string): string {
  return `${trainPathById(tid)}/${GROUP_SCANNER_CARD}`;
}

/** Per-train ClearBox subgroup — `Extensions/ClearBox/<train_id>/`. */
export function clearboxPathById(tid: string): string {
  return `${GROUP_CLEARBOX}/${tid}`;
}

export function clearboxPath(index: number): string {
  return clearboxPathById(trainId(index));
}

export function correctionDataPathById(tid: string): string {
  return `${clearboxPathById(tid)}/${DS_CORRECTION_DATA}`;
}

export function inverseCorrectionDataPathById(tid: string): string {
  return `${clearboxPathById(tid)}/${DS_INVERSE_CORRECTION_DATA}`;
}

export function scanFieldCorrectionFilePath(index: number): string {
  return `${trainPath(index)}/${DS_SCAN_FIELD_CORRECTION_FILE}`;
}

File: rust/src/capabilities/v1_1/layout.rs
//! File_Version 1.1 on-disk HDF5 layout.
//!
//! Deliberately independent of `capabilities/v1_0/layout.rs` — nothing here
//! imports from or refers to v1.0, so v1.0 can change or be removed later
//! without affecting v1.1. See `docs/migrations/v1_0_to_v1_1.md` for the full
//! migration manifest this layout implements.

pub const FILE_VERSION: &str = "1.1";

pub const ROOT_MACHINE: &str = "Machine";
pub const ROOT_OPTICAL_TRAINS: &str = "Machine/Optical_Trains";
pub const ROOT_EXTENSIONS: &str = "Extensions";
pub const GROUP_CLEARBOX: &str = "Extensions/ClearBox";
pub const ROOT_OPCUA: &str = "Extensions/TM_OPCUA";
pub const OPCUA_CLIENT: &str = "Extensions/TM_OPCUA/Client";
pub const OPCUA_PIPE: &str = "Extensions/TM_OPCUA/Pipe";
pub const OPCUA_TRIGGERS: &str = "Extensions/TM_OPCUA/Triggers";

pub const TRAIN_ID_PREFIX: &str = "Optical_Train_";
pub const GROUP_SCANNER: &str = "Scanner";
pub const GROUP_LIGHT_SOURCE: &str = "Light_Source";
pub const GROUP_COLLIMATOR: &str = "Collimator";
pub const GROUP_SCANNER_CARD: &str = "Scanner_Card";
pub const GROUP_POWER_CHARACTERIZATION: &str = "Power_Characterization";
pub const DS_DERIVATION_EQUATION_CONSTANTS: &str = "Derivation_Equation_Constants";
pub const DS_CHARACTERIZATION_POINTS: &str = "Characterization_Points";
pub const DS_CORRECTION_DATA: &str = "Correction_Data";
pub const DS_INVERSE_CORRECTION_DATA: &str = "Inverse_Correction_Data";
pub const DS_SCAN_FIELD_CORRECTION_FILE: &str = "scan_field_correction_file";

pub fn train_id(index: usize) -> String {
    format!("{TRAIN_ID_PREFIX}{:02}", index + 1)
}

pub fn train_path_by_id(tid: &str) -> String {
    format!("{ROOT_OPTICAL_TRAINS}/{tid}")
}

pub fn scanner_path_by_id(tid: &str) -> String {
    format!("{}/{GROUP_SCANNER}", train_path_by_id(tid))
}

pub fn light_source_path_by_id(tid: &str) -> String {
    format!("{}/{GROUP_LIGHT_SOURCE}", train_path_by_id(tid))
}

pub fn collimator_path_by_id(tid: &str) -> String {
    format!("{}/{GROUP_COLLIMATOR}", train_path_by_id(tid))
}

pub fn scanner_card_path_by_id(tid: &str) -> String {
    format!("{}/{GROUP_SCANNER_CARD}", train_path_by_id(tid))
}

pub fn clearbox_path_by_id(tid: &str) -> String {
    format!("{GROUP_CLEARBOX}/{tid}")
}

pub fn correction_data_path_by_id(tid: &str) -> String {
    format!("{}/{DS_CORRECTION_DATA}", clearbox_path_by_id(tid))
}

pub fn inverse_correction_data_path_by_id(tid: &str) -> String {
    format!("{}/{DS_INVERSE_CORRECTION_DATA}", clearbox_path_by_id(tid))
}

pub fn scan_field_correction_file_path_by_id(tid: &str) -> String {
    format!("{}/{DS_SCAN_FIELD_CORRECTION_FILE}", train_path_by_id(tid))
}
